### ISSUE TITLE
[Client bug]: Intermittent authentication issue when executing Group's root site call

### ISSUE THREAD

**Issue by lballaire** [microsoftgraph/msgraph-sdk-dotnet#1232]
**Describe the bug**
We noticed that we are seeing a an error coming back from the Graph SDK when using a specific query pretty often and it does not seem to happen on other queries. Note that the class handling the query is used by multiple paths in our codebase since the query gets the SharePoint root site of a Group. For context, we have a multi-tenant app that interacts with multiple workloads using Graph.
 
The error we get is "invalid audience Uri" : 

```
Status Code: Unauthorized
Microsoft.Graph.ServiceException: Code: unauthenticated
Message: Invalid audience Uri 'https://graph.microsoft.com/'.
Inner error:
   Code: invalidAudienceUri
 at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
  at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
  at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
  at Microsoft.Graph.SiteRequest.GetAsync(CancellationToken cancellationToken)
```

**To Reproduce**
It seems intermittent, but only happens on a specific query which is implemented like so : 

```
var groupSitesRequest = client.Groups[groupId.ToString()]
   .Sites["root"]
   .Request()
   .Select(QuerySelectHelper.FormatSelectParameters(GroupRootSiteQueryParameters));
 
var result = await groupSitesRequest.GetAsync();
```

We have added some validation to ensure that it is not a race condition in our code. The added log does not show up and it still only happens on that query. We also verified when providing an actual token of the wrong audience and the error is different.

**Expected behavior**
Since the token is valid, the call should succeed.

**Client version**
4.14.0

**Additional context**
We have a multi-tenant application with around 50 different queries. Only the one provided in the "To Reproduce" section does this and the logic to obtain the token is the same for all our queries. Therefore, we suspect it might be an authentication issue on the SDK or the API itself when transitioning between the Groups workload to SharePoint since this query technically should use both.

To help investigate, here are timestamps and the client-request-id of a few errors : 
Jan 25 8:44:11 EST - 7a3a1aa0-1052-1000-4ddb-f84edf6b1799
Jan 25 8:28:48 EST - 98391aa0-c0fd-c000-ae06-48b8636ea74e
Jan 20 0:31:20 EST - 4a8218a0-b00e-3000-5f85-92eac20f49c2

**Comment by andrueastman** [comment]
Hey @lballaire,
Thanks for raising this. 

To help us understand this better, are you able to confirm if this is happening to only single tenants or multiple tenants? Could you share request-ids from multiple tenants?

> We also verified when providing an actual token of the wrong audience and the error is different.

What exact error do you get in this scenario?

**Comment by lballaire** [comment]
> To help us understand this better, are you able to confirm if this is happening to only single tenants or multiple tenants?

Multiple Tenants. We're running a multi-tenant SaaS Application. The dotnet process will run multiple Graph Api SDK queries concurrently for different tenants at the same time.

> Could you share request-ids from multiple tenants?

I did at the bottom of my initial comment, but I can provide more up-to-date ones. Here are some example from our test environment using some lab tenants.

```
Tenant 4638eb7e-5f42-47f7-ac48-ae7abbaf14b4
Group 278cfbbb-73e8-4f7d-b649-36f264af303e
date: 2022-02-21T13:07:40
request-id: a300da8e-03e3-45be-95a0-2bc9890b208a
client-request-id: a300da8e-03e3-45be-95a0-2bc9890b208a

Tenant 8c34a947-07b7-45bc-bb23-758b95a34e93
Group beb30e4a-5b9a-4c23-9082-66724038101a
date: 2022-02-21T13:07:45
request-id: 51bd6ba7-f4cd-4cd4-92a6-dd5791081d17
client-request-id: 51bd6ba7-f4cd-4cd4-92a6-dd5791081d17

Tenant 69feac3b-980a-4103-a0fb-39e39768bbc0
Group ba691daa-9dea-4a80-bbf1-85537dc69fac
date: 2022-02-21T13:20:43
request-id: b245ef4a-74d5-4197-92ef-be6ea9201db4
client-request-id: b245ef4a-74d5-4197-92ef-be6ea9201db4
```

> What exact error do you get in this scenario?

When  providing the wrong token, we get this error instead which is a little different.

```
Microsoft.Graph.ServiceException
Code: InvalidAuthenticationToken
Message: Access token validation failure. Invalid audience.
Inner error:
	AdditionalData:
	date: 2022-02-21T13:17:11
	request-id: 94aac592-0164-48e2-9d29-06060b9cce79
	client-request-id: 94aac592-0164-48e2-9d29-06060b9cce79
ClientRequestId: 94aac592-0164-48e2-9d29-06060b9cce79

   at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.SiteRequest.GetAsync(CancellationToken cancellationToken)
```

We thought it was an error on our side at first due to the nature of our multi tenant application but we ruled that out because the error message we get is different.

**Comment by andrueastman** [comment]
Thanks for the extra information @lballaire.

> We thought it was an error on our side at first due to the nature of our multi-tenant application but we ruled that out because the error message we get is different

That makes sense. From the graph logs it looks like the workload is throwing this error and not the graph service as in the case for the wrong token. I have raised an issue with them to find out if they can clear up what could be going wrong with the request. 

In the meantime, by any chance could you share what authentication provider/mechanism you are using to acquire the token?
It might also be helpful to take a token that results in the failing request and paste it over at [jwt.ms](jwt.ms) and share that result as well for the workload team as well.

**Comment by lballaire** [comment]
Thank you for the follow-up @andrueastman.

We currently use ADAL to obtain the token with our own implementation of a token cache to encrypt/decrypt tokens. That being said, we are currently moving towards MSAL since ADAL is being deprecated. We do not have mechanisms that give us access to the token that is used for request except locally while debugging. Since this is intermittent, it might be difficult to reproduce locally.

To authenticate the request with the SDK, we use a DelegatedAuthenticationProvider that fetches from the token cache. We ruled out an issue with this part of our code too since we only see failures with that specific call and, as mentioned initially, we have around 50 other queries to Graph.

I am not sure if this answers your question. Otherwise, what specific information would help with your investigation?

**Comment by andrueastman** [comment]
Thanks for the extra context @lballaire.

Any chance you could grab one of the failing tokens and paste it over at [jwt.ms](https://jwt.ms) and share the result?
One thing we would need to confirm is if the token format is okay as ADAL is known to return the [old token forma](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-differences-adal-net#v10-and-v20-tokens)t that is might cause issues with Microsoft graph.

**Comment by andrueastman** [comment]
Hey @lballaire,

I have received feedback from the relevant workload team. 
It seems that the failure is caused by tenantId present in the token does not match the tenantId for which the request is being made. This therefore causes the above failure. 

The solution suggested is to ensure that the tenantId is explicitly passed to the authentication library so that it gives back a token with the correct tenantId as failing to do so will result in the library incorrectly inferring the tenantId.

Any chance you are able to confirm if this can be done?

**Comment by lballaire** [comment]
Hi @andrueastman,

Thank you very much for the follow-up. We are going to look into this further and give you an update around the beginning of next week.

In one of your previous messages, you mentioned issues with ADAL and the old token format. Considering we are starting to move towards MSAL, what are the chances that this issue might be addressed by doing so?

**Comment by andrueastman** [comment]
I believe we should be able to resolve this with MSAL.NET and get assistance/feedback with any issues that may crop up as it is currently actively supported. 

For now, as advised we would start by making sure we explicitly pass the `tenantId` to the auth lib. 
One of the ways is using the `WithTenantId` modifier as documented [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-initializing-client-applications#modifiers-common-to-public-and-confidential-client-applications)

**Comment by lballaire** [comment]
The example you provided about `WithTenantId` is only applicable to MSAL.NET. Unfortunately, as I mentioned previously, we are starting to migrate to MSAL, but still have a lot of work ahead of us.

I cannot seem to find documentation about providing the `TenantId` to ADAL. What is currently in production with the issue is the following simplified : 

```
var authContext = new AuthenticationContext($"https://login.microsoftonline.com/{tenantId}", tokenCache);
var tokenResult = await authenticationContext.AcquireTokenAsync(resourceId, clientAssertionCertificate);
return tokenResult.AccessToken;
```

When obtaining a token successfully, `tokenResult.TenantId` is always `null` even though we did provide a `tenantId` in the authority. There does not seem to be another signature that allows us to provide a `tenantId` with the `Microsoft.IdentityModel.Clients.ActiveDirectory` library.

However, when looking at the claims of `tokenResult.AccessToken`, I see that the `tid` claim is assigned and this is what is sent to Graph when making a request. Could you confirm that the `tenantId` that did not match was from the `tid` claim?

**Comment by andrueastman** [comment]
Hey @lballaire,

Apologies for the delayed response. 

Yes. That is correct. The feedback from the SharePoint team was that the `tenantId` that did not match was from the `tid` claim in the accessToken.

**Comment by lballaire** [comment]
Hi @andrueastman,

We added an internal mechanism in an isolated test environment to retrieve a token in case of failure. In every case, the `tid` claim did match the actual `tenantId` on which the request was executed. Here are 3 stripped down versions of token claims with the associated `TenantId` and `ClientRequestId` : 

```
TenantId: 679e56c2-2faf-459b-b730-d52fdf229839
ClientRequestId: de8b557c-8873-43af-b4d4-90739ab7e489
Timestamp: 2022-03-01T19:58:21.6973843+00:00
{
  "aud": "https://graph.microsoft.com/",
  "iss": "https://sts.windows.net/679e56c2-2faf-459b-b730-d52fdf229839/",
  "idp": "https://sts.windows.net/679e56c2-2faf-459b-b730-d52fdf229839/",
  "roles": [
    "Group.Read.All",
    "Directory.ReadWrite.All",
    "Sites.Read.All",
    "Group.ReadWrite.All",
    "TeamsAppInstallation.ReadWriteForUser.All",
    "Mail.Send"
  ],
  "tid": "679e56c2-2faf-459b-b730-d52fdf229839",
}

TenantId: 4638eb7e-5f42-47f7-ac48-ae7abbaf14b4
ClientRequestId: cf26a8b9-e8dd-4c1c-ac22-4f88687aaa31
Timestamp: 2022-03-01T19:48:57.665647+00:00
{
    "aud": "https://graph.microsoft.com/",
    "iss": "https://sts.windows.net/4638eb7e-5f42-47f7-ac48-ae7abbaf14b4/",
    "idp": "https://sts.windows.net/4638eb7e-5f42-47f7-ac48-ae7abbaf14b4/",
    "roles": [
      "Group.Read.All",
      "Directory.ReadWrite.All",
      "Sites.Read.All",
      "Group.ReadWrite.All",
      "TeamsAppInstallation.ReadWriteForUser.All",
      "Mail.Send"
    ],
    "tid": "4638eb7e-5f42-47f7-ac48-ae7abbaf14b4"
  }


TenantId: 149fde9a-edc8-472d-96b4-27ce727b5728
ClientRequestId: 61b61759-71cf-4b05-98f7-fac9554ea08e
Timestamp: 2022-03-01T19:48:20.4188904+00:00
{
    "aud": "https://graph.microsoft.com/",
    "iss": "https://sts.windows.net/149fde9a-edc8-472d-96b4-27ce727b5728/",
    "idp": "https://sts.windows.net/149fde9a-edc8-472d-96b4-27ce727b5728/",
    "roles": [
        "Group.Read.All",
        "Directory.ReadWrite.All",
        "Sites.Read.All",
        "Group.ReadWrite.All",
        "TeamsAppInstallation.ReadWriteForUser.All",
        "Mail.Send"
    ],
    "tid": "149fde9a-edc8-472d-96b4-27ce727b5728"
} 
```

Like I mentioned earlier, we have over 50 different requests which we send maybe a hundreds per second and the fact that only this request gives us this error indicates that the problem is most probably not on our side. If the problem was that the token is not on the right tenant, wouldn't the Graph service return the unauthorized response before even reaching the SharePoint workload?

**Comment by andrueastman** [comment]
Hi @lballaire,

Thanks again for this. Let me feed this back to the workload team and circle back with what they give back.

**Comment by lballaire** [comment]
Hi @andrueastman,

I searched for the `tenantId` you provided and since it is a tenant that belongs to us and consented our app, I started to dig a little bit more. 

I managed to reproduce the issue by executing a request on an existing group while forcing the use of a token from another tenant. That way, I get the same error that we see in our production environment. This could very well indicate that the issue is in our code and that we have a potential race condition when getting the token or it could be somewhere in the Graph SDK. 

I know I mentioned previously that we have a lot of other requests and only the one to obtain the root site of a group fails. However, I found out that the other requests to not behave the same way. For example : 
- `/groups/<groupId>/sites/root` with wrong token : 401 with "invalid audience uri"
- `/groups/<groupId>` with wrong token : 404

We will try to add more logging over the week to pinpoint where the issue could be located. I'll let you know if it seems to be in our code or the SDK.

**Comment by lballaire** [comment]
Hi @andrueastman,

Sorry for the delay. After a few changes on our side, the issue does not seem to be as present as much. I will close the issue. 

Thank you for your help.

**Comment by paulj1010** [comment]
Hi @lballaire - Quick question, how did you resolve your issue? We have exactly the same problem on that `/groups/xxx/sites/root` endpoint (same app based access token works fine for say the `/groups/xxx` but fails with Invalid audience on that particular endpoint). Thanks.

**Comment by lballaire** [comment]
Hey @paulj1010

What we did is a bit tricky, but seems to be successful so far. Basically, we replaced the `IHttpProvider` by a custom implementation in order to return a singleton `HttpClient` for every request.

We took the [implementation in the SDK](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dev/src/Microsoft.Graph.Core/Requests/HttpProvider.cs), but injected the `HttpClient` directly instead of letting it instantiate it.

```
public CustomClientGraphHttpProvider(HttpClient httpClient)
{
    this._httpClient = httpClient;
}
```

The HttpClient itself has retry policies for some status codes, but not on unauthorized responses.

At the same time, we found ourselves adding some cache on the result of some calls where the result does not change much like this one. This minimizes the number of times the requests are executed, but the error rate dropped before rolling out the cache, so we are pretty confident that the singleton HttpClient fixed the issue.

Hope this helps.

**Comment by lballaire** [comment]
Hey @paulj1010 - Sorry if I wasn't clear. Actually the `HttpClient` does not retry on unauthorized responses. Normally, if you aren't authorized once, you shouldn't be afterwards.

We think it worked because we suspect a conflict where the instance that is provided by the SDK does not authenticate properly when supplying only an `HttpProvider`. We have added logging to pinpoint the issue and noticed that some requests were made to another tenant when they shouldn't.

One of the reasons why we took this approach is because we needed to add retry handlers to transient errors (timeouts, 503 that are not throttling, etc.). This was the only way where we could alter the actual retry handlers coming from Graph and it seems to have fixed our issue at the same time. It also allowed us to use an HttpClient from the .NET DI stack. We were inspired by [this issue](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/318) and [this one](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/202).
